### ISSUE TITLE
Check the `SocketClosed` event work

### ISSUE THREAD

**Issue by OlegRa** [alpacahq/alpaca-trade-api-csharp#403]
@OlegRa I've done some testing with the `SocketClosed` handler and the `DataStreamingClient .WithReconnect` method. I don't believe they are working correctly. 

Test 1 - with the following code

```cs
DataStreamingClient = marketData.Environment
    .GetAlpacaDataStreamingClient(new SecretKey(marketData.API_KEY, marketData.API_SECRET))
    .WithReconnect(new ReconnectionParameters
    {
        MaxReconnectionAttempts = 10
    });
```
I connect to the streaming client and subscribe for ~50 tickers.  I then disconnect my wifi. In my mind, this configuration should cause re-connection attempts.  I let it sit for 15 minutes and in no case did I get an indication that it attempted to perform a re-connection nor did `DataStreamingClient.SocketClosed` handler gets called. 

Test 2 - with the following code
 
```cs
DataStreamingClient = marketData.Environment
    .GetAlpacaDataStreamingClient(new SecretKey(marketData.API_KEY, marketData.API_SECRET));
```

I connect to the streaming client and subscribe for ~50 tickers.  I then disconnect my wifi. In my mind, this configuration should cause re-connection attempts.  I let it sit for 15 minutes and in no case did my `DataStreamingClient.SocketClosed` handler get called. 

Please let me know if you want me to open a separate bug for this.

_Originally posted by @Tifoid in https://github.com/alpacahq/alpaca-trade-api-csharp/issues/398#issuecomment-950394828_

**Comment by Tifoid** [comment]
@OlgaRa ... I cant confirm that this change is working fully ... but I will detail the results of my testing 

Test 1 
with  MaxReconnectionAttempts count set to 10  
DataStreamingClient OnError handler invoked when disconnected wifi 

Test 2
with out  MaxReconnectionAttempts reconnection logic enabled  
DataStreamingClient 

- [ ] OnSocketClosed handler invoked when disconnected wifi 

I am not certain why the disconnections from the wife resulted in differing handler invocation ... perhaps that is what you were intending. 

That all said, I am unable to properly test if the  MaxReconnectionAttempts logic is working correctly as I do not have a way to force a disconnect from just the Alpaca web service.  What I can confirm is that since adding the  MaxReconnectionAttempts logic I have consistently gotten reliable minute data every day for the past week.

**Comment by OlegRa** [comment]
@Tifoid If you use an automatic reconnection wrapper the `OnSocketClosed` event will be raised only after `MaxReconnectionAttempts` unsuccessful attempts. Each disconnect will be reported via the `OnError` event.

I use the TCPView tool from SysInternals for dropping the specific connection under Windows. AFAIR the Netstat allows you to do the same under Linux.

**Comment by Tifoid** [comment]
@OlegRa Using the TCPView tool I tested a scenario where I had the  MaxReconnectionAttempts = 10. I then proceeded to close the connection (via TCPView) in excess of 10 times. Each time my process appears to have reconnected (my OnConnected handler was invoked with a status of Authorized). What was strange is that even with MaxReconnectionAttempts = 10 I was able to close the connection more than 10 times and it never invoked my OnSocketClosed or OnError handlers. This struck me as odd.

**Comment by OlegRa** [comment]
The `OnSocketClosed` will be raised if reconnection logic will not able to reconnect 10 times. If you just drop the connection and application reconnects immediately you'll see only the `OnConnected` events. The `OnError` event will not be reported if you drop connection via TCPView because in this case, you have a "recoverable" socket error. Right now, you cannot control the list of these "recoverable" errors, but maybe I'll add the ability to configure them in the next releases.

**Comment by OlegRa** [comment]
@Tifoid Next week I'll close this issue with the "resolved" status due to inactivity. If you still have problems with this event - just tell me or re-open it if you'll experiencing it again later.

**Comment by Tifoid** [comment]
I'm good with you closing this issue.  I'll keep tracking and testing.  If I run into additional problems I'll create a new issue.  thanks!
